internal/validation/errors_test.go

```diff
@@ -0,0 +1,470 @@
+	err := newError(doc, field, assert.AnError)
+	t.Log(err)
+	require.Error(t, err)
+	require.Contains(t, err.Error(), fmt.Sprintf("validating SliceErrorTestDoc.primitiveSliceSlice[1][1]: %s", assert.AnError))
+}
+
+// Tests for maps
+
+func TestNewValidationErrorPrimitiveMap(t *testing.T) {
+	st := &MapErrorTestDoc{
+		PrimitiveMap: map[string]string{
+			"abc": "def",
+			"ghi": "jkl",
+		},
+	}
+
+	doc, field := references(t, st, &st.PrimitiveMap, "ghi")
+	err := newError(doc, field, assert.AnError)
+	t.Log(err)
+	require.Error(t, err)
+	require.Contains(t, err.Error(), fmt.Sprintf("validating MapErrorTestDoc.primitiveMap[\"ghi\"]: %s", assert.AnError))
+}
+
+func TestNewValidationErrorStructPointerMap(t *testing.T) {
+	st := &MapErrorTestDoc{
+		StructPointerMap: map[string]*ErrorTestDoc{
+			"abc": {
+				ExportedField: "abc",
+				OtherField:    123,
+			},
+			"ghi": {
+				ExportedField: "ghi",
+				OtherField:    456,
+			},
+		},
+	}
+
+	doc, field := references(t, st, &st.StructPointerMap["ghi"].OtherField, "")
+	err := newError(doc, field, assert.AnError)
+	t.Log(err)
+	require.Error(t, err)
+	require.Contains(t, err.Error(), fmt.Sprintf("validating MapErrorTestDoc.structPointerMap[\"ghi\"].otherField: %s", assert.AnError))
+}
+
+func TestNewValidationErrorNestedPrimitiveMap(t *testing.T) {
+	st := &MapErrorTestDoc{
+		NestedPointerMap: map[string]*map[string]string{
+			"abc": {
+				"def": "ghi",
+			},
+			"jkl": {
+				"mno": "pqr",
+			},
+		},
+	}
+
+	doc, field := references(t, st, st.NestedPointerMap["jkl"], "mno")
+	err := newError(doc, field, assert.AnError)
+	t.Log(err)
+	require.Error(t, err)
+	require.Contains(t, err.Error(), fmt.Sprintf("validating MapErrorTestDoc.nestedPointerMap[\"jkl\"][\"mno\"]: %s", assert.AnError))
+}
+
+// Special cases
+
+func TestNewValidationErrorTopLevelIsNeedle(t *testing.T) {
+	st := &ErrorTestDoc{
+		ExportedField: "abc",
+		OtherField:    42,
+	}
+
+	doc, field := references(t, st, st, "")
+	err := newError(doc, field, assert.AnError)
+	require.Error(t, err)
+	require.Contains(t, err.Error(), fmt.Sprintf("validating ErrorTestDoc: %s", assert.AnError))
+}
+
+func TestNewValidationErrorUntaggedField(t *testing.T) {
+	st := &ErrorTestDoc{
+		ExportedField: "abc",
+		OtherField:    42,
+		NoTagField:    123,
+	}
+
+	doc, field := references(t, st, &st.NoTagField, "")
+	err := newError(doc, field, assert.AnError)
+	require.Error(t, err)
+	require.Contains(t, err.Error(), fmt.Sprintf("validating ErrorTestDoc.NoTagField: %s", assert.AnError))
+}
+
+func TestNewValidationErrorOnlyYamlTaggedField(t *testing.T) {
+	st := &ErrorTestDoc{
+		ExportedField: "abc",
+		OtherField:    42,
+		NoTagField:    123,
+		OnlyYamlKey:   "abc",
+	}
+
+	doc, field := references(t, st, &st.OnlyYamlKey, "")
+	err := newError(doc, field, assert.AnError)
+	require.Error(t, err)
+	require.Contains(t, err.Error(), fmt.Sprintf("validating ErrorTestDoc.onlyYamlKey: %s", assert.AnError))
+}
+
+type ErrorTestDoc struct {
```

Comment:
nit: these testing types don't need to be exported, only their field names have to be